src/main/kotlin/org/rust/cargo/CargoProjectDescription.kt

```diff
@@ -1,40 +1,47 @@
 package org.rust.cargo
 
-import com.intellij.util.PathUtil
-import org.rust.cargo.commands.impl.CargoMetadata
+import com.intellij.util.containers.BidirectionalMap
+import com.intellij.util.containers.MultiMap
 import java.io.File
-import java.io.Serializable
-import java.util.*
 
+
+/**
+ * Rust project model represented roughly in the same way as in Cargo itself.
+ *
+ * [org.rust.cargo.toolchain.CargoMetadataService] is responsible for providing a [CargoProjectDescription] for
+ * an IDEA module.
+ */
 class CargoProjectDescription private constructor(
-    val packages: Collection<Package>,
-    private val rootModule: Package
+    private val idToPackage: BidirectionalMap<Int, Package>,
+    private val dependencies: MultiMap<Int, Int>
 ) {
+    val packages: Collection<Package> get() = idToPackage.values
+    val rawDependencies: MultiMap<Int, Int> get() = MultiMap(dependencies)
 
     class Package(
```

Comment:
Yep, you are right! I've struggled quite a bit with the design of this `ProjectDescription` class, but I am still not happy :(

The project is a DAG of packages. Serializing graphs is awkward, and we have **two** formats for serializing this thing. One is `cargo metadata` output, which has snake_case names and other peculiarities (`target_kind`). Another is IDEAs XML format, which requires all fields to be mutable and nullable.